### ISSUE TITLE
[3.18.3] Python Protobuf caused segmentation fault only on Mac

### ISSUE THREAD

**Issue by jcwchen** [protocolbuffers/protobuf#10691]
**What version of protobuf and what language are you using?**
Version: 3.18.3
Language: Python

**What operating system (Linux, Windows, ...) and version?**
Mac. Windows and Linux worked normally

**What runtime / compiler are you using (e.g., python version or gcc version)**
Python 3.7-3.10

**What did you do?**
1. Install protoc (3.18.3 or others)
2. `pip install protobuf==3.18.3`
3. `git clone  --recurse-submodules https://github.com/onnx/onnx.git && cd onnx`
4. `python setup.py install`

**What did you expect to see**
Install onnx package successfully without crash, or at least a meaningful error

**What did you see instead?**

```
--   ONNX_WERROR               : ON
--   ONNX_BUILD_TESTS          : ON
--   ONNX_BUILD_BENCHMARKS     : OFF
-- 
--   Protobuf compiler         : /Users/runner/work/1/s/protobuf/protobuf_install/bin/protoc
--   Protobuf includes         : /Users/runner/work/1/s/protobuf/protobuf_install/include
--   Protobuf libraries        : /Users/runner/work/1/s/protobuf/protobuf_install/lib/libprotobufd.a
--   BUILD_ONNX_PYTHON         : ON
--     Python version        : 
--     Python executable     : /Users/runner/hostedtoolcache/Python/3.8.14/x64/bin/python
--     Python includes       : /Users/runner/hostedtoolcache/Python/3.8.14/x64/include/python3.8
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/runner/work/1/s/.setuptools-cmake-build
[  2%] Running gen_proto.py on onnx/onnx.in.proto
[  2%] Creating directories for 'googletest'
[  3%] Performing download step (git clone) for 'googletest'
Processing /Users/runner/work/1/s/onnx/onnx.in.proto
Writing /Users/runner/work/1/s/.setuptools-cmake-build/onnx/onnx_ONNX_NAMESPACE_FOO_BAR_FOR_CI-ml.proto
Writing /Users/runner/work/1/s/.setuptools-cmake-build/onnx/onnx_ONNX_NAMESPACE_FOO_BAR_FOR_CI-ml.proto3
Writing /Users/runner/work/1/s/.setuptools-cmake-build/onnx/onnx-ml.pb.h
generating /Users/runner/work/1/s/.setuptools-cmake-build/onnx/onnx_pb.py
Cloning into 'googletest'...
[  4%] Running C++ protocol buffer compiler on /Users/runner/work/1/s/.setuptools-cmake-build/onnx/onnx_ONNX_NAMESPACE_FOO_BAR_FOR_CI-ml.proto
/Users/runner/work/1/s/.setuptools-cmake-build/tools/protoc-gen-mypy.sh: line 4:  8355 Segmentation fault: 11  "/Users/runner/hostedtoolcache/Python/3.8.14/x64/bin/python" -u "$DIR"/protoc-gen-mypy.py
--mypy_out: protoc-gen-mypy: Plugin failed with status code 139.
make[2]: *** [onnx/onnx_ONNX_NAMESPACE_FOO_BAR_FOR_CI-ml.pb.cc] Error 1
make[1]: *** [CMakeFiles/gen_onnx_proto.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Anything else we should know about your project / environment**
I only see this kind of Segmentation fault on Mac with Python Protobuf==3.18.3 specifically. Other versions such as previous 3.18, 3.19, 3.20 all worked well. I guess probably something is broken in Python Protobuf 3.18.3 wheel for Mac? Python Protobuf 3.18.3 is just released recently (9/13), and I didn't see any related issue reported from the web. I tried a few Mac machines and all encountered this issue. Since it's a seg fault so it's quite hard to debug. Any suggestion will be appreciated. Thank you.

**Comment by jcwchen** [comment]
Hi @zhangskz,
Any update on this? Please let me know if there is anything I can help from my side. Thank you for looking into this issue.

**Comment by zhangskz** [comment]
Hi jcwchen@, I spent some time trying to repro this but had some setup issues with onnx. I ended up following the instructions from https://github.com/onnx/onnx#mac but am still hitting an issue with "Protobuf compiler version doesn't match library version 3.18.3"  (though my protoc --version is indeed 3.18.3). 

Can you help with a few things to make this easier for us to repro?
1. Can you confirm if this issue reproduces in 3.19.5, 3.20.2, or 21.6? 3.18.x has technically fallen out of our support window, so actually moving to a newer version (ideally 21.x) would be better anyways.
2. If you use our pre-built protoc (https://github.com/protocolbuffers/protobuf/releases/tag/v3.18.3) instead of pip install, do you still hit the same issue?
3. If you can provide a repro for this issue using protobuf directly without onnx, this will make it easier for us to dig into this.

**Comment by jcwchen** [comment]
Thank you for the investigation. I think it's more like a Python Protobuf issue instead of protoc issue, because it only happened with Python Protobuf 3.18.3 regardless the version of protoc.

> Can you confirm if this issue reproduces in 3.19.5, 3.20.2, or 21.6? 3.18.x has technically fallen out of our support window, so actually moving to a newer version (ideally 21.x) would be better anyways.

Python Protobuf 3.19.5, 3.19.6, 3.20.2, 3.20.3, 4.21.6 all work well from my Mac. Only Python Protobuf 3.18.3 will bump into a seg fault crash issue.

> If you use our pre-built protoc (https://github.com/protocolbuffers/protobuf/releases/tag/v3.18.3) instead of pip install, do you still hit the same issue?

I tried protoc 3.18.3 from your link and it will still bump into issue if the environment has Python Protobuf 3.18.3. Other Python Protobuf version work fine.

> If you can provide a repro for this issue using protobuf directly without onnx, this will make it easier for us to dig into this.

Yes, but it's quite hard for me to decouple a small piece from the original onnx build. I will try, but meanwhile I am happy to help you resolve onnx build issue if any.

> "Protobuf compiler version doesn't match library version 3.18.3" (though my protoc --version is indeed 3.18.3).

Have you `git clean -xdf` under onnx directory? Or, a quick workaround might be finding a clean Mac environment with single installed protoc. Protoc version shouldn't matter here so you can have any version >=3.16. As long as onnx can find your protoc and you use Python Protobuf 3.18.3, you should be able to repro this crash. Feel free to let me know if you see other onnx build issue. Thank you for your time.

**Comment by jcwchen** [comment]
Typically ONNX ecosystem does not frequently bump dependency version (Protobuf, NumPy, and etc.) because ONNX downstream tools might need lower version for different kinds of reasons. Therefore, ideally I would like to keep required Protobuf version for ONNX as flexible as possible.

3.18.3 is a fresh patch release very recently (a month ago) and this is the only version seems broken with some use cases on Mac. According to the experiments, there might be something needs to be further fixed in the patch updates (commits from 3.18.2 to 3.18.3) for Mac. Although Python Protobuf 3.18.0-3.18.2 are fine, going forward we cannot really use them due to https://github.com/protocolbuffers/protobuf/security/advisories/GHSA-8gq9-2x98-w8hf. I was just worried other users might also see similar issues on Mac with Python Protobuf 3.18.3 if they also need this new patch release.

**Comment by zhangskz** [comment]
Note that since 3.18.x has exited our support window, [GHSA-h4h5-3hr4-j3g2](https://github.com/protocolbuffers/protobuf/security/advisories/GHSA-h4h5-3hr4-j3g2) (Java advisory) was not patched in 3.18.x and this/future issues will similarly not be patched in this branch either. It is still recommended to move off of 3.18 (at least to 3.19) for this reason.
